[WIP] profiles: remove obsolete package.accept_keywords entries

andrey-utkin opened the pull request:
The removed entries are now keyworded for both arm and arm64 in main
Gentoo tree.

Bash script:

    cat profiles/targets/rpi3/package.accept_keywords/* \
    | grep -v -E '^$|^#|^=' \
    | awk '{ print $1 }' \
    | sort -u \
    | while read CATPN; do
          REPO=/usr/portage
          if grep -q 'KEYWORDS=.*arm .*arm64' $REPO/$CATPN/*.ebuild &>/dev/null; then
              echo $CATPN
          fi
      done \
    | while read CATPN; do
          # How the heck to not match anything which just begins with CATPN and has some suffix?
          sed -i -e "\\:^${CATPN}:d" profiles/targets/rpi3/package.accept_keywords/*
      done \
      ;

    find profiles/targets/rpi3/package.accept_keywords/* -empty | xargs git rm -f

sakaki- commented:
Thanks, been meaning to do that for some time ^-^

Unfortunately, as I'm travelling atm, I can't immediately check the impact of this.

Would you mind therefore getting a current copy of the gentoo-on-rpi3-64bit image, applying this PR, then doing a pretend emptytree @world emerge (with bdeps) to ensure nothing wants to be downgraded (or worse!) as a result?

If that works out OK, I'll merge it.

Best, sakaki 
-- 
sakaki@deciban.com

andrey-utkin commented:
Will take me a while to get back to this, sorry, don't be surprised.

Commit:
profiles: don't use wildcard in accept_keywords for packages keyworded for arm64

Most of entries in profiles/targets/rpi3/package.accept_keywords/*
specify value "* ~*", which means "package is visible if it is in stable
or testing status on any architecture".

This means, once any affected package gets a new ebuild having *any*
keyword, e.g. ~amd64, it becomes "visible" as a pending update for the
system. This is not a bad trick per se, especially for a deliberate
"bleeding edge" system, and especially in earlier times when arm64
didn't maintain a stable keyword. But this means potentially dealing
with more cases of breakage, because Gentoo staff hasn't yet assured
that the new version works on arm64 architecture.

This commit turns such lines, for packages which are keyworded for
arm64 (at least in unstable status, ~arm64), into lines with just the
package name, which means "package is visible if it is in stable or
testing status on *host* architecture". This means arm64 in this case.
It might lead to some confusion for people who are trying to use the
overlay for arm (32-bit) systems, but they don't seem to be the target
audience anyway. I am open to changing this commit to list arm64
explicitly.

The script used to produce the change:

    REPO=/usr/portage
    cat profiles/targets/rpi3/package.accept_keywords/* \
    | grep -v -E '^$|^#|^=' \
    | awk '{ print $1 }' \
    | sort -u \
    | while read CATPN; do
          if grep -q 'KEYWORDS=.*arm64' $REPO/$CATPN/*.ebuild &>/dev/null; then
    	  # Keyworded, stable or unstable
    	  # Replace '$CAT/$PN * ~*' which is 'accept stable or testing on any arch'
    	  # with '$CAT/$PN' which is 'accept testing on host arch'
    	  sed -i -e "s:^${CATPN}[ ]\+[*] [~][*]$:${CATPN}:" profiles/targets/rpi3/package.accept_keywords/*
          fi
      done

andrey-utkin commented:
I reviewed what I was attempting to do here, and realized that my previous approach was too far from correctness.

I have run `emerge --pretend --emptytree --with-bdeps=y @world`, and got this result:
```
Total: 1122 packages (1 downgrade, 1 in new slot, 1120 reinstalls, 1118 binaries), Size of downloads: 670,296 KiB

WARNING: One or more updates/rebuilds have been skipped due to a dependency conflict:

media-sound/pulseaudio:0

  (media-sound/pulseaudio-13.0:0/0::gentoo, ebuild scheduled for merge) conflicts with
    >=media-sound/pulseaudio-12.2-r3[rpi-deglitch(-)] required by (dev-embedded/rpi-64bit-meta-1.5.1-r7:0/0::genpi64, binar
y scheduled for merge)


media-video/ffmpeg:0

  (media-video/ffmpeg-4.2.1:0/56.58.58::gentoo, ebuild scheduled for merge) conflicts with
    >=media-video/ffmpeg-4.2-r1[v4l2m2m-fix(-)] required by (dev-embedded/rpi-64bit-meta-1.5.1-r7:0/0::genpi64, binary sche
duled for merge)


sys-boot/rpi3-64bit-firmware:0

  (sys-boot/rpi3-64bit-firmware-1.20190925:0/0::genpi64, ebuild scheduled for merge) conflicts with
    ~sys-boot/rpi3-64bit-firmware-1.20190819[-dtbo(+)] required by (sys-kernel/bcm2711-kernel-bis-bin-4.19.73.20190924:0/0:
:genpi64, binary scheduled for merge)
    ^                             ^^^^^^^^^^
    ~sys-boot/rpi3-64bit-firmware-1.20190819[-dtbo(+)] required by (sys-kernel/bcmrpi3-kernel-bis-bin-4.19.73.20190924:0/0:
:genpi64, binary scheduled for merge)
    ^                             ^^^^^^^^^^
```
The only downgrade is media-gfx/gimp, and it is explanable - the package has "just recently" got keyworded arm64, but not in the latest ebuilds unfortunately.

I am now re-running having added `--usepkg=n --getbinpkg=n`.

andrey-utkin commented:
I see the same conflicts if i run the same command without my patch. So perhaps exactly that command is not indicative?
